### PR TITLE
[#245] Query replication slot for WAL streaming starting point

### PR DESCRIPTION
**Pull request by l32zhao**
Implement the feature to query the replication slot using the READ_REPLICATION_SLOT command from the PostgreSQL replication protocol. This command should be issued before attempting to use IDENTIFY_SYSTEM as a last resort when the archive directory is empty.

### CODE REVIEW

**Review by jesperpedersen — CHANGES_REQUESTED** [review]

File: src/libpgmoneta/wal.c [hunk 1]
```diff
@@ -211,28 +212,31 @@ pgmoneta_wal(int srv, char** argv)
    }
    config->servers[srv].cur_timeline = cur_timeline;
 
-   wal_find_streaming_start(d, &timeline, &high32, &low32, segsize);
-   if (timeline == 0)
+   if (wal_read_replication_slot(ssl, socket, config->servers[srv].wal_slot, &high32, &low32, &timeline, segsize))
```
[the remark this answers is not in the document]
Comment (l32zhao) [reply src/libpgmoneta/wal.c]: Done with the latest squashed commit

**Comment by l32zhao** [comment]
# Changes
1. Error Label: Added an error label (error:) for centralized error handling.
2. PostgreSQL 15+ Check: Introduced read_replication variable to check for PostgreSQL 15+ and fallback if necessary.
3. Local Variables for Timeline and LSN: Assigned timeline and LSN to local variables before assigning to output parameters.
4. Improved Error Logging: Enhanced error messages to include the server name for better traceability.

File: src/libpgmoneta/wal.c [hunk 2]
```diff
@@ -211,28 +213,36 @@ pgmoneta_wal(int srv, char** argv)
    }
    config->servers[srv].cur_timeline = cur_timeline;
 
-   wal_find_streaming_start(d, &timeline, &high32, &low32, segsize);
-   if (timeline == 0)
+   if (wal_read_replication_slot(ssl, socket, config->servers[srv].wal_slot, config->servers[srv].name, &high32, &low32, &timeline, segsize))
```
Comment (jesperpedersen) [inline src/libpgmoneta/wal.c:216]: We should not execute the command if PostgreSQL < 15

**Review by jesperpedersen — CHANGES_REQUESTED** [review]

File: src/libpgmoneta/wal.c [hunk 3]
```diff
@@ -211,28 +213,41 @@ pgmoneta_wal(int srv, char** argv)
...
 
-      if (xlogpos == NULL)
+   if (read_replication)
+   {
+      if (wal_read_replication_slot(ssl, socket, config->servers[srv].wal_slot, config->servers[srv].name, &high32, &low32, &timeline, segsize))
       {
-         goto error;
+         read_replication = 0;   // Fallback if not PostgreSQL 15+
       }
+   }
```
Comment (l32zhao) [inline src/libpgmoneta/wal.c:224]: Check PostgreSQL version before read replication slot

**Comment by jesperpedersen** [comment]
```
????C????IDENTIFY_SYSTEM?Z????I
540000005E0003736C6F745F747970650000000000000000000019FFFFFFFFFF
FF0000726573746172745F6C736E0000000000000000000019FFFFFFFFFFFF00
00726573746172745F746C6900000000000000000000140008FFFFFFFF000044
0000001A000300000008706879736963616CFFFFFFFFFFFFFFFF430000001A52
4541445F5245504C49434154494F4E5F534C4F54005A0000000549
T???^??slot_type????????????????
???restart_lsn??????????????????
?restart_tli???????????????????D
??????????physical????????C????R
EAD_REPLICATION_SLOT?Z????I
2024-05-25 13:15:58 ERROR wal.c:657 Error occurred when executing READ_REPLICATION_SLOT for slot repl on server primary
```

```
psql postgres 
psql (16.3)
Type "help" for help.

postgres=# \q
```

File: src/libpgmoneta/wal.c [hunk 4]
```diff
@@ -211,28 +213,41 @@ pgmoneta_wal(int srv, char** argv)
...
-   if (timeline == 0)
-   {
-      // use current xlogpos as last resort
-      timeline = cur_timeline;
-      xlogpos_size = strlen(pgmoneta_query_response_get_data(identify_system_response, 2)) + 1;
-      xlogpos = (char*)malloc(xlogpos_size);
+   read_replication = (config->servers[srv].version >= 15) ? 1 : 0;
 
-      if (xlogpos == NULL)
+   if (read_replication)
```
Comment (Jubilee101) [inline src/libpgmoneta/wal.c:218]: Remember this is only the last resort. Normally we get our starting point by finding the latest wal segment in the wal archive directory, say `backup/primary/wal`. So you should read replication slot only if that archive directory is empty, meaning `wal_find_streaming_start` gets a 0 timeline.

File: src/libpgmoneta/wal.c [hunk 5]
```diff
@@ -625,6 +641,59 @@ pgmoneta_wal(int srv, char** argv)
...
+   lsn = pgmoneta_query_response_get_data(read_slot_response, 1);
+   if (wal_convert_xlogpos(lsn, high32, low32, segsize))
+   {
+      pgmoneta_log_error("Failed to convert LSN from replication slot %s on server %s", slot, name);
+      goto error;
+   }
+
+   *timeline = local_timeline;
+
+   pgmoneta_log_info("Success read replica: timeline %d, LSN %s", *timeline, lsn);
```
Comment (jesperpedersen) [inline src/libpgmoneta/wal.c:684]: This is debug, and replication slot - not replica

**Review by jesperpedersen — CHANGES_REQUESTED** [review]

File: src/libpgmoneta/wal.c (hunk 5 above)
Comment (l32zhao) [reply src/libpgmoneta/wal.c]: Sorry, this is a dirty log, and I've removed it.

**Comment by jesperpedersen** [comment]
Rebased, and merged.

Thanks for your contribution !

**Pull request closed without merging**